File: ros2_ws/src/video_publisher_py/video_publisher_py/publisher.py
import os
import sys

import cv2
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge


class Publisher(Node):
    def __init__(self, video_path: str):
        super().__init__('video_publisher')

        if not os.path.exists(video_path):
            self.get_logger().error(f"No such file: {video_path}")
            rclpy.shutdown()
            return

        self.bridge = CvBridge()
        self.video = cv2.VideoCapture(video_path)
        if not self.video.isOpened():
            self.get_logger().error(f"Can't open video: {video_path}")
            rclpy.shutdown()
            return
        
        fps = self.video.get(cv2.CAP_PROP_FPS)
        if fps <= 0:
            fps = 30.0
        
        self.publisher = self.create_publisher(Image, 'camera/image', 10)

        timer_period = 1.0/fps
        self.timer = self.create_timer(timer_period, self.timer_callback)

    def timer_callback(self):
        ret, frame = self.video.read()
        if not ret:
            self.video.set(cv2.CAP_PROP_POS_FRAMES, 0) #Reset vide when ends
            return

        msg = self.bridge.cv2_to_imgmsg(frame, encoding='bgr8')
        msg.header.stamp = self.get_clock().now().to_msg()

        self.publisher.publish(msg)


def main(args=None):
    rclpy.init(args=args)

    if len(sys.argv) < 2:
        print("Usage")
        print("ros2 run video_publisher_py talker <video_path>")
        return

    node = Publisher(video_path=sys.argv[1])
    rclpy.spin(node)

    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ros2_ws/src/video_publisher/src/publisher.cpp
#include <sstream>
#include <string>

#include "cv_bridge/cv_bridge.hpp"
#include "image_transport/image_transport.hpp"
#include "opencv2/core/mat.hpp"
#include "opencv2/highgui.hpp"
#include "opencv2/videoio.hpp"
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/header.hpp"
#include "sensor_msgs/msg/image.hpp"

int main(int argc, char ** argv)
{
  if (argc < 2) {
    std::cerr << "Uso: ros2 run video_publisher talker <ruta_video>\n";
    return 1;
  }

  rclcpp::init(argc, argv);
  auto node = rclcpp::Node::make_shared("video_publisher");

  image_transport::ImageTransport it(node);
  auto pub = it.advertise("camera/image", 1);

  std::string video_path = argv[1];

  cv::VideoCapture cap(video_path);

  if (!cap.isOpened()) {
    std::cerr << "No se pudo abrir el video: " << video_path << "\n";
    return 1;
  }

  double fps = cap.get(cv::CAP_PROP_FPS);
  if (fps <= 0.0) fps = 30.0;

  rclcpp::WallRate loop_rate(fps);

  cv::Mat frame;
  std_msgs::msg::Header header;

  auto logger = rclcpp::get_logger("video_publisher");
  RCLCPP_INFO(logger, "Publishing video");
  while (rclcpp::ok()) {

    cap >> frame;

    // Si llega al final del video, terminar
    if (frame.empty()){
      RCLCPP_INFO(logger, "Finished publishing video");
      break;
    }

    header.stamp = node->now();

    auto msg = cv_bridge::CvImage(header, "bgr8", frame).toImageMsg();

    pub.publish(msg);

    rclcpp::spin_some(node);
    loop_rate.sleep();
  }

  rclcpp::shutdown();
  return 0;
}